### ISSUE TITLE
[FEATURE] Being able to connect with mac address instead of IP

### ISSUE THREAD

**Issue by Mozoloa** [bitfocus/companion-module-behringer-xair#43]
As you guys may have experienced already, when you can't give a fixed ip address to the XR/MR consoles on a local network, the IP tends to jump around from time to time depending on how the stars are aligned.

This has been a big problem for some of my clients as it kills all the buttons for the streamdeck and they're stuck until I remote into their computer and change the IP (they can't ever remember how to do it). Right now I've made a script they can click that searches through `arp -a` for the console mac address (that's stored in a txt file somewhere) then finds the ip and replaces it in companion's db and restart the app, but this is very clunky and `arp -a` is unreliable.

I wondered if we could try to add a Mac Address option for connexion, with maybe a little function that scans the local network the same way X Air Edit does to find the IP (no idea how tho, but they seem to scan for `68-27-19-**-**-**` machines somehow)

**Comment by Mozoloa** [comment]
Bumping this issue, as we're moving to companion 3.0, I believe such a functionality has been done already on the X32 module for a while. It'd be great if it could store the XR mac address and always reconnect to it no matter the local IP

**Comment by istnv** [comment]
This is one of the reasons I try never to use a dynamic IP on show-critical devices.
... however ...
The _mac_ address is not easily found in the libraries used.
The name could be used to re-locate a unit.

**Comment by Mozoloa** [comment]
I've tried with the name but sometimes windows simply can't find it.

My installations are pretty specific, they're in content creators/streamer setups and those can change a lot, and they sometimes move and/or change network providers etc etc and it's always going to be a problem at some point. Also I don't always have access to the admin panel of their local network.

The way I do it in powershell, so native windows stuff, is as follows (can probably be improved but it could give you idea)

1. - Find the subnet and ping every address once in parallel
2. - Do` arp -a ` (`arp -a` won't display local devices unless your machine has interacted with it, hence the mass ping)
3. - All XR/MR network interfaces I've met had a mac adress that started by one of those: 
 ```
68-27-19
00-1e-c0
04-91-62
e8-eb-1b
```
4. - So you can fetch the line with those address and you have the IP of the console(s)
5. - From there you can /xinfo and let the user decide which one is the right one if there are multiple consoles and save the mac address somewhere

Now whenever the IP drops out you can just do 1 and 2 and use the saved mac address to find the exact IP

**Comment by istnv** [comment]
The internal mixer name is not DNS related so you can't ping by name.
You could ping the broadcast address (x.x.x.255). 
The XAir will respond and show up in the arp table.
`/xinfo` to the broadcast address is how XAir Edit locates units.

**Comment by Mozoloa** [comment]
That's some very useful info for my solution until something is done on the module, thanks a lot !

**Comment by Julusian** [comment]
> I believe such a functionality has been done already on the X32 module for a while

That is not the case. What the X32 does do is uses the builtin discovery supported by the X32 to list the discovered devices, but once you select one it remembers that specific ip address.

Maybe the xair supports the same form of discovery that could be used to help find it after the ip address change?
I figured out how to do it for the x32 with X32edit which would auto-discover them. The implementation is a self contained class https://github.com/bitfocus/companion-module-behringer-x32/blob/master/src/device-detector.ts

**Comment by istnv** [comment]
@Julusian The XAir units (as well as the X32) respond to an OSC `/xinfo` request sent to the broadcast address. The return info includes the mixer's IP and Name. I was going to store the mixer 'name' argument to locate it again if the IP changed.

**Comment by Mozoloa** [comment]
Tried yesterday with some basic osc commands for my solution, broadcasting /xinfo on the local network works like a charm

**Comment by istnv** [comment]
@Mozoloa I added a new PR that should auto re-locate the unit.

**Comment by Mozoloa** [comment]
that's great, pretty hyped to refresh my solution with all those upgrades

**Comment by istnv** [comment]
Feature added in latest release

**Comment by Mozoloa** [comment]
Thanks a lot, I haven't had the time to check, will it then automatically reconnect upon local ip change ?

**Comment by istnv** [comment]
@Mozoloa UDP The module starts the variable request sync and if there is no response after 10 seconds, it will look to see if the a unit with the same Mixer Name is on the network. If it is located, it will change the IP to match and re-start automatically.

**Comment by Mozoloa** [comment]
That's great, thanks for all you do ! cheerz
